simplicity-sys/src/alloc.rs

```diff
@@ -0,0 +1,104 @@
+use std::alloc::{self, Layout};
+use std::mem;
+
+/// A type whose alignment is greater equal that of any C type.
+/// Allocations with this alignment are valid for all C types.
+/// Remake of `max_align_t` from `stddef.h`.
+/// [16 bytes is the greatest alignment of any architecture Rust currently supports](https://github.com/rust-lang/rust/blob/61223975d46f794466efa832bc7562b9707ecc46/library/std/src/sys/pal/common/alloc.rs).
+#[repr(align(16))]
+#[derive(Default, Copy, Clone)]
+pub struct AlignedType([u8; 16]);
+
+/// Minimal alignment which is valid for all C types.
+pub const MIN_ALIGN: usize = mem::align_of::<AlignedType>();
+
+/// Signature of an allocator function.
+type AllocatorFn = unsafe fn(Layout) -> *mut u8;
+
+/// Allocate `size_bytes` many bytes using the `allocator` function
+/// and return a pointer.
+///
+/// # Safety
+///
+/// `allocator` must be [`alloc::alloc`] or [`alloc::alloc_zeroed`].
+///
+/// Allocated bytes must be freed using [`rust_free`].
+unsafe fn allocate(size_bytes: usize, allocator: AllocatorFn) -> *mut u8 {
+    // Allocate MIN_ALIGN + size_bytes many bytes
+    // Panic if too many bytes are tried to be allocated
+    let size_prefixed_bytes = MIN_ALIGN + size_bytes;
+    let layout = Layout::from_size_align(size_prefixed_bytes, MIN_ALIGN).unwrap();
+    let ptr_prefix = allocator(layout);
+    if ptr_prefix.is_null() {
+        alloc::handle_alloc_error(layout);
+    }
+    // Write the number of allocated bytes to memory
+    (ptr_prefix as *mut usize).write(size_prefixed_bytes);
+    // Return a pointer to the size_bytes many allocated bytes behind the counter
+    // We need to offset the pointer by MIN_ALIGN to keep alignment
+    // This means there is a gap of MIN_ALIGN - sizeof(size_t) many unused bytes
+    // We asserted that MIN_ALIGN >= sizeof(size_t), so this gap is nonnegative
```

Comment:
In 22c8741c74c73fab7588d969ec2d21fc6d3b95c3

Curiously, in rust-secp we actually *do* assert this, but here we only comment it. (I see that there's a unit test about this, but we should also put an assertion into `allocate` itself.)